### ISSUE TITLE
False unreachable_code positive with `do catch` and early return

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust#43336]
```rust
#![feature(catch_expr, never_type)]

struct Error {
    cause: &'static str,
}

fn oops() -> Result<(), Error> {
    Err(Error { cause: "epic fail" } )
}

fn main() {
    let Err(Error { cause }): Result<!, Error> = do catch {
        oops()?;
        return; // OK
    };
    eprintln!("Nuclear meltdown initiated: {}", cause);
    return ();
}
```

```
warning: unreachable statement
  --> src/main.rs:16:5
   |
16 |     eprintln!("Nuclear meltdown initiated: {}", cause);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unreachable_code)] on by default
   = note: this error originates in a macro outside of the current crate

    Finished dev [unoptimized + debuginfo] target(s) in 0.86 secs
     Running `target/debug/playground`
Nuclear meltdown initiated: epic fail
```

**Comment by crumblingstatue** [comment]
This seems to be fixed on current nightly (45fba43b3 2018-02-10). There is no warning about an unreachable statement anymore.

**Comment by crumblingstatue** [comment]
Since the syntax has changed to try, and now the code doesn't even compile, I'll just close this issue
